### ISSUE TITLE
About TDI

### ISSUE THREAD

**Issue by SeregaKR** [calloatti/TBZ#1]
Hello.
For compatibility reasons I have to work with old FoxPro app. I want to freshen it up a bit so I decided to add tabbed interface using your class TDI. But I got stuck testing the example.
When I run main.prg from FoxPro IDE, or when I run exe (again from IDE) just after building it, everything seems fine. But when I try to run exe directly from Windows, the form shows up and closes after several seconds.
I tried adding messageboxes to find the problem but I got stuck. The last working moment I found was in windowproc function in tdibar (Case m.pMsg = WM_SHOWWINDOW).

Can you help please? Sorrry for offtop.

**Comment by calloatti** [comment]
No problem. This is how to use TDI:

Drop a tdimain control into a top level form. Let's call it mainform

Add code in mainform.tdimain.tdiinint that calls a form that has buttons that allow the user to call the forms of your application.

This menu form an all the other forms should be showwindow = In Top Level Form.

Now your problem is not a TDI problem, but a VFP problem. You are missing a "Read Events" call

Read Events "pauses" the execution of code and the executable waits there for user actions until a clear events is issued

So basically your main entry point to your project should be a prg that calls your main form and then issues a read events like this:

```
*!* main.prg

_InitErrorHandler()

If _IsAppRunning() = .T. And Application.StartMode = 4 Then
	Cancel
Endif

Do ctl32.prg
Do _base.prg
Do palm.prg

_SetPath()

Do Form Main

If Application.StartMode = 0 Then
	Return
Else
	On Shutdown Clear Events
	Read Events
Endif

On Error
On Shutdown
Cancel
Return
```

Then on mainform.unload you have:

`Clear Events`

That is all.

**Comment by calloatti** [comment]
https://web.archive.org/web/20151105154631/http://www.ctl32.com:80/tdi/tdi.html

**Comment by SeregaKR** [comment]
Thank you. I knew I was missing something but I was not sure what exactly it was.

**Comment by Digital-Fisher** [comment]
Greetings, I would like to be able to download the Ctl32 TDI Tabbed control but I have already done an exhaustive search on the net and I can't find it. can you upload it somewhere to download it? I still develop and maintain applications in vfp

**Comment by SeregaKR** [comment]
> Greetings, I would like to be able to download the Ctl32 TDI Tabbed control but I have already done an exhaustive search on the net and I can't find it. can you upload it somewhere to download it? I still develop and maintain applications in vfp

Hello. I can't find my download from the URL you mentioned, but I found another link.
http://ctl32.blogspot.com/p/ctl32-downloads.html  look for [TDI Alpha class](https://drive.google.com/uc?export=download&id=1yNNXGz_Q4ME9esfX7CvPi_SAO21VbRCC)
